README.md
# FacturAPI

[![npm version](https://badge.fury.io/js/facturapi-es6.svg)](https://badge.fury.io/js/facturapi-es6)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/Flet/semistandard)

This is the unofficial Node.js wrapper for https://www.facturapi.io

FacturAPI makes it easy for developers to generate valid Invoices in Mexico (known as Factura Electrónica or CFDI).

If you've ever used [Stripe](https://stripe.com) or [Conekta](https://conekta.io), you'll find FacturAPI very straightforward to understand and integrate in your server app.

## Install

```bash
npm install --save facturapi-es6
```

## Getting started

### Authenticate with your API Key

Make sure you have created your free account on [FacturAPI](https://www.facturapi.io) and that you have your **API Keys**.

```typescript
import { Facturapi } from "facturapi-es6";
const facturapi = new Facturapi("YOUR_API_KEY");
```

### Create a customer

```typescript
facturapi.customers
  .create({
    legal_name: "Walter White", // Razón social
    tax_id: "WIWA761018", // RFC
    email: "walterwhite@gmail.com", // Optional but useful to send invoice by email
    address: {
      street: "Av. de los Rosales",
      exterior: "123",
      neighborhood: "Tepito",
      zip: "06800"
      // city, municipality and state are filled automatically from the zip code
      // but if you want to, you can override their values
      // city: 'México',
      // municipality: 'Cuauhtémoc',
      // state: 'Ciudad de México'
    }
  })
  .then(customer => {
    // Remember to store the customer.id in your records.
    // You will need it to create an invoice for this customer.
  })
  .catch(err => console.log(err)); // Handle the error.
```

### Create a product

```typescript
facturapi.products
  .create({
    product_key: "4319150114", // Clave Producto/Servicio from SAT's catalog. Log in to FacturAPI and use our tool to look it up.
    description: "Apple iPhone 8",
    price: 20000 // price in MXN.
    // By default, taxes are calculated from the price with IVA 16%
    // But again, you can override that by explicitly providing a taxes array
    // taxes: [
    //   { type: Facturapi.TaxType.IVA, rate: 0.16 },
    //   { type: Facturapi.TaxType.ISR, rate: 0.03666, withholding: true }
    // ]
  })
  .then(product => {
    // Remember to store the product.id in your records.
    // You will need it to create an invoice for this product.
  })
  .catch(err => console.log(err)); // Handle the error.
```

### Create an invoice

```typescript
facturapi.invoices.create({
  customer: 'YOUR_CUSTOMER_ID',
  payment_form: Facturapi.PaymentForm.TRANSFERENCIA_ELECTRONICA, // Constant from SAT's catalog. Check out our documentation to learn more.
  items: [{
    quantity: 1, // Optional. Defaults to 1.
    product: 'YOUR_PRODUCT_ID' // You can also pass a product object instead
  }] // Add as many products as you want to include in your invoice
}).then(invoice => { ... });
```

#### Download your invoice

```typescript
// Once you have successfully created your invoice, you can...
const fs = require('fs');
facturapi.invoices.downloadZip(invoice.id) // or downloadPdf or downloadXml
  .then(zipStream => {
    // stream containing the PDF and XML as a ZIP file
    // Save your invoice to a folder
    const myZipFile = fs.createWriteStream('/path/to/destination/folder');
    zipStream.pipe(myZipFile);
    myZipFile.on('finish', () => {
      // Finished downloading, Yay!
    });
```

#### Send your invoice by email

```typescript
// Send the invoice to your customer's email (if any)
facturapi.invoices
  .sendByEmail(invioce.id) // Also returns a Promise
  .then(() => {
    // Successfully sent
  })
  .catch(err => console.log(err)); // Handle the error.
```

## Documentation

There's more you can do with this library: List, retrieve, update, and remove Customers, Products and Invoices.

Visit the full documentation at http://docs.facturapi.io.

## Help

### Found a bug?

Please report it on the Issue Tracker

### Want to contribute?

Send us your PR! We appreciate your help :)

### Contact us!

contacto@facturapi.io


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.2.1] 2019-11-09

- add support to use in es6 project (typescript)

```typescript
import { Facturapi } from "facturapi-es6";
const facturapi = new Facturapi("YOUR_API_KEY");
```

for build need install typescript global and run `tsc` in project directory this generate a `dist` folder

## [2.0.0] 2018-08-04

## Breaking changes

- Now you must create the Facturapi instance using the `new` keyword every time.

**Before:**

```javascript
// This was allowed
const facturapi = Facturapi("YOUR_API_KEY");
```

**Now:**

```javascript
// Now you must always use new
const facturapi = new Facturapi("YOUR_API_KEY");
```

## [1.2.0] 2018-08-04

### Fixed

- Reject with an Error, not with an object

## [1.1.0] 2018-05-06

### Added

- Support Organizations API

## [1.0.0] - 2018-05-01

### Added

- Search `product_key`s using `facturapi.products.keys('your search')`
- Search `unit_key`s using `facturapi.products.units('your search')`
- Constants for PaymentMethod, InvoiceType, InvoiceUse, InvoiceRelation

### Breaking changes

- Now contants should be accessed as static properties from the Facturapi class, instead of from the instance.

**Before:**

```javascript
const facturapi = new Facturapi("YOUR_API_KEY");
console.log(facturapi.TaxType.IVA); // > IVA
```

**Now:**

```javascript
console.log(Facturapi.TaxType.IVA); // > IVA
```

## [0.1.3] - 2017-06-20

### Fixed

- Protocol should be HTTPS

## [0.1.2] - 2017-06-20

### Added

- First release
- Wrapper methods for:
  - Customers
  - Products
  - Invoices
- Added README file


src/types/enums.ts
export namespace enums {
  export enum PaymentForm {
    EFECTIVO = "01",
    CHEQUE_NOMINATIVO = "02",
    TRANSFERENCIA_ELECTRONICA = "03",
    TARJETA_DE_CREDITO = "04",
    MONEDERO_ELECTRONICO = "05",
    DINERO_ELECTRONICO = "06",
    VALES_DE_DESPENSA = "08",
    DACION_EN_PAGO = "12",
    SUBROGACION = "13",
    CONSIGNACION = "14",
    CONDONACION = "15",
    COMPENSACION = "17",
    NOVACION = "23",
    CONFUSION = "24",
    REMISION_DE_DEUDA = "25",
    PRESCRIPCION_O_CADUCIDAD = "26",
    A_SATISFACCION_DEL_ACREEDOR = "27",
    TARJETA_DE_DEBITO = "28",
    TARJETA_DE_SERVICIOS = "29",
    POR_DEFINIR = "99"
  }
  export type PaymentFormList = [
    { value: "01"; label: "Efectivo" },
    { value: "02"; label: "Cheque nominativo" },
    { value: "03"; label: "Transferencia electrónica de fondos" },
    { value: "04"; label: "Tarjeta de crédito" },
    { value: "05"; label: "Monedero electrónico" },
    { value: "06"; label: "Dinero electrónico" },
    { value: "08"; label: "Vales de despensa" },
    { value: "12"; label: "Dación en pago" },
    { value: "13"; label: "Pago por subrogación" },
    { value: "14"; label: "Pago por consignación" },
    { value: "15"; label: "Condonación" },
    { value: "17"; label: "Compensación" },
    { value: "23"; label: "Novación" },
    { value: "24"; label: "Confusión" },
    { value: "25"; label: "Remisión de deuda" },
    { value: "26"; label: "Prescripción o caducidad" },
    { value: "27"; label: "A satisfacción del acreedor" },
    { value: "28"; label: "Tarjeta de débito" },
    { value: "29"; label: "Tarjeta de servicios" },
    { value: "99"; label: "Por definir" }
  ];

  export enum TaxType {
    IVA = "IVA",
    IEPS = "IEPS",
    ISR = "ISR"
  }

  export enum PaymentMethod {
    PAGO_EN_UNA_EXHIBICION = "PUE",
    PAGO_EN_PARCIALIDADES_DIFERIDO = "PPD"
  }
  export enum InvoiceUse {
    ADQUISICION_MERCANCIAS = "G01",
    DEVOLUCIONES_DESCUENTOS_BONIFICACIONES = "G02",
    GASTOS_EN_GENERAL = "G03",
    CONSTRUCCIONES = "I01",
    MOBILIARIO_Y_EQUIPO_DE_OFICINA = "I02",
    EQUIPO_DE_TRANSPORTE = "I03",
    EQUIPO_DE_COMPUTO = "I04",
    DADOS_TROQUELES_HERRAMENTAL = "I05",
    COMUNICACIONES_TELEFONICAS = "I06",
    COMUNICACIONES_SATELITALES = "I07",
    OTRA_MAQUINARIA = "I08",
    HONORARIOS_MEDICOS = "D01",
    GASTOS_MEDICOS_POR_INCAPACIDAD = "D02",
    GASTOS_FUNERALES = "D03",
    DONATIVOS = "D04",
    INTERESES_POR_CREDITOS_HIPOTECARIOS = "D05",
    APORTACIONES_VOLUNTARIAS_SAR = "D06",
    PRIMA_SEGUROS_GASTOS_MEDICOS = "D07",
    GASTOS_TRANSPORTACION_ESCOLAR = "D08",
    CUENTAS_AHORRO_PENSIONES = "D09",
    SERVICIOS_EDUCATIVOS = "D10",
    POR_DEFINIR = "P01"
  }

  export enum InvoiceType {
    INGRESO = "I",
    EGRESO = "E",
    TRASLADO = "T",
    NOMINA = "N",
    PAGO = "P"
  }

  export enum InvoiceRelation {
    NOTA_DE_CREDITO = "01",
    NOTA_DE_DEBITO = "02",
    DELOVUCION_DE_MERCANCIA = "03",
    SUSTITUCION_DE_CFDI_PREVIOS = "04",
    TRASLADOS_DE_MERCANCIA_FACTURADOS_PREVIAMENTE = "05",
    FACTURA_POR_TRASLADOS_PREVIOS = "06",
    APLICACION_DE_ANTICIPO = "07",
    PAGOS_EN_PARCIALIDADES = "08",
    PAGOS_DIFERIDOS = "09"
  }

  export enum TaxSystem {
    GENERAL_LEY_DE_PERSONAS_MORALES = "601",
    PERSONAS_MORALES_CON_FINES_NO_LUCRATIVOS = "603",
    SUELDOS_Y_SALARIOS = "605",
    ARRENDAMIENTO = "606",
    DEMAS_INGRESOS = "608",
    CONSOLIDACION = "609",
    RESIDENTES_EN_EL_EXTRANJERO = "610",
    INGRESOS_POR_DIVIDENDOS_SOCIOS_Y_ACCIONISTAS = "611",
    PERSONAS_FISICAS_CON_ACTIVIDADES_EMPRESARIALES_Y_PROFESIONALES = "612",
    INGRESOS_POR_INTERESES = "614",
    SIN_OBLIGACIONES_FISCALES = "616",
    SOCIEDADES_COOPERATIVAS_DE_PRODUCCION = "620",
    REGIMEN_DE_INCORPORACION_FISCAL = "621",
    ACTIVIDADES_AGRICOLAS_GANADERAS_SILVICOLAS_Y_PESQUERAS = "622",
    OPCIONAL_PARA_GRUPOS_DE_SOCIEDADES = "623",
    COORDINADOS = "624",
    HIDROCARBUROS = "628",
    REGIMEN_DE_ENAJENACION_O_ADQUISICION_DE_BIENES = "607",
    PREFERENTES_Y_EMPRESAS_MULTINACIONALES = "629",
    ENAJENACION_DE_ACCIONES_EN_BOLSA_DE_VALORES = "630",
    REGIMEN_DE_LOS_INGRESOS_POR_OBTENCION_DE_PREMIOS = "615"
  }
  export enum cancelationStatus {
    NO_CANCELADO = "none",
    CANCELACION_PENDIENTE = "pending",
    CANCELACION_ACEPTADA = "accepted",
    CANCELACION_RECHAZADA = "rejected",
    SOLICITUD_EXPIRADA = "expired"
  }
}
